src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.ConnectionManager.cs

```diff
@@ -0,0 +1,161 @@
+                _pools = new ConcurrentDictionary<string, ConcurrentQueue<JsonRpcConnection>>(concurrencyLevel: 4, capacity: 4);
+
+                _enableConnectionPool = enableConnectionPool;
+                _shutdownLock = new ReaderWriterLockSlim(LockRecursionPolicy.NoRecursion);
+            }
+
+            public Task<Connection> TryCreateConnectionAsync(string serviceName, object callbackTarget, CancellationToken cancellationToken)
+            {
+                // pool is not enabled by option
+                if (!_enableConnectionPool)
+                {
+                    // RemoteHost is allowed to be restarted by IRemoteHostClientService.RequestNewRemoteHostAsync
+                    // when that happens, existing Connection will keep working until they get disposed.
+                    //
+                    // now question is when someone calls RemoteHostClient.TryGetConnection for the client that got
+                    // shutdown, whether it should gracefully handle that request or fail after shutdown.
+                    // for current expected usage case where new remoteHost is only created when new solution is added,
+                    // we should be fine on failing after shutdown.
+                    //
+                    // but, at some point, if we want to support RemoteHost being restarted at any random point, 
+                    // we need to revisit this to support such case by creating new temporary connections.
+                    // for now, I dropped it since it felt over-designing when there is no usage case for that yet.
+                    return TryCreateNewConnectionAsync(serviceName, callbackTarget, cancellationToken);
+                }
+
+                // when callbackTarget is given, we can't share/pool connection since callbackTarget attaches a state to connection.
+                // so connection is only valid for that specific callbackTarget. it is up to the caller to keep connection open
+                // if he wants to reuse same connection
+                if (callbackTarget != null)
+                {
+                    return TryCreateNewConnectionAsync(serviceName, callbackTarget, cancellationToken);
+                }
+
+                return TryGetConnectionFromPoolAsync(serviceName, cancellationToken);
+            }
+
+            private async Task<Connection> TryGetConnectionFromPoolAsync(string serviceName, CancellationToken cancellationToken)
+            {
+                var queue = _pools.GetOrAdd(serviceName, _ => new ConcurrentQueue<JsonRpcConnection>());
+                if (queue.TryDequeue(out var connection))
+                {
+                    return new PooledConnection(this, serviceName, connection);
+                }
+
+                return new PooledConnection(this, serviceName, (JsonRpcConnection)await TryCreateNewConnectionAsync(serviceName, callbackTarget: null, cancellationToken).ConfigureAwait(false));
```

Comment:
❗️ This code needs to handle `TryCreateNewConnectionAsync` returning null.

Reply:
thank you!